definitions/npm/react-redux_v4.x.x/flow_v0.30.x-/react-redux_v4.x.x.js

```diff
@@ -1,78 +1,86 @@
-  mapDispatchToProps: (dispatch: Dispatch, ownProps: $Diff<P, MP>) => DP,
-  mergeProps: (stateProps: SP, dispatchProps: DP, ownProps: $Diff<P, MP>) => MP,
-  options?: {pure?: boolean, withRef?: boolean}
-) => (component: (props: P) => any) => Class<React$Component<void, $Diff<P, MP>, void>>;
-
-type ConnectNoState = <D, P, S, C: React$Component<D, P, S>, DP, Dispatch: Function>(
-    mapStateToProps: null | void,
-    mapDispatchToProps: (dispatch: Dispatch, ownProps: $Diff<$Diff<P, DP>, D>) => DP,
-    mergeProps: null | void,
-    options?: {pure?: boolean, withRef?: boolean}
-  ) => (component: Class<C>) => Class<React$Component<D, $Diff<P, DP>, S>>;
-
-type ConnectNoStateStatless = <P, DP, Dispatch: Function>(
-    mapStateToProps: null | void,
-    mapDispatchToProps: (dispatch: Dispatch, ownProps: $Diff<P, DP>) => DP,
-    mergeProps: null | void,
-    options?: {pure?: boolean, withRef?: boolean}
-  ) => (component: (props: P) => any) => Class<React$Component<void, $Diff<P, DP>, void>>;
-
-type ConnectDispatch = <D, P, S, C: React$Component<D, P, S>, SP, Dispatch: Function>(
-  mapStateToProps: (state: Object, ownProps: $Diff<$Diff<$Diff<P, {dispatch: Dispatch}>, SP>, D>) => SP,
-  mapDispatchToProps: null | void,
-  mergeProps: null | void,
-  options?: {pure?: boolean, withRef?: boolean}
-) => (component: Class<C>) => Class<React$Component<D, $Diff<$Diff<P, {dispatch: Dispatch}>, SP>, S>>;
-
-type ConnectDispatchStateless = <P, SP, Dispatch: Function>(
-  mapStateToProps: (state: Object, ownProps: $Diff<$Diff<P, {dispatch: Dispatch}>, SP>) => SP,
-  mapDispatchToProps: null | void,
-  mergeProps: null | void,
-  options?: {pure?: boolean, withRef?: boolean}
-) => (component: (props: P) => any) => Class<React$Component<void, $Diff<$Diff<P, {dispatch: Dispatch}>, SP>, void>>;
-
-type ConnectDefault = <D, P, S, C: React$Component<D, P, S>, Dispatch: Function>() =>
-  (component: Class<C>) => Class<React$Component<D, $Diff<P, {dispatch: Dispatch}>, S>>;
-
-type ConnectDefaultStateless = () =>
-  <P>(component: (props: P) => any) => Class<React$Component<void, $Diff<P, {dispatch: Function}>, void>>;
+import type { Dispatch, Store } from 'redux'
 
 declare module 'react-redux' {
-  declare var exports: {
-    connect: ConnectAll
-      & ConnectAllStateless
-      & ConnectMerged
-      & ConnectMergedStateless
-      & ConnectNoState
-      & ConnectNoStateStatless
-      & ConnectDispatch
-      & ConnectDispatchStateless
-      & ConnectDefault
-      & ConnectDefaultStateless;
-    Provider: ReactClass<{store: Object, children?: any}>;
+
+  /*
+
+    S = State
+    A = Action
+    OP = OwnProps
+    SP = StateProps
+    DP = DispatchProps
+
+  */
+
+  declare type MapStateToProps<S, OP: Object, SP: Object> = (state: S, ownProps: OP) => SP | MapStateToProps<S, OP, SP>;
+
+  declare type MapDispatchToProps<A, OP: Object, DP: Object> = ((dispatch: Dispatch<A>, ownProps: OP) => DP) | DP;
```

Comment:
This case doesn't work correctly in practice. When this argument is not a function, the types don't match up. (DP here can be just an object of actions, but the type of the same props in the component are all functions.